### DISCUSSION TITLE
[Key_Shared] How to ensure the order when unack and send msg concurrent?

### DISCUSSION THREAD

**Discussion by xrayw** [apache/pulsar#21863] [Q&A] [answered]
How does Key Shared mode ensure the order of consume when consumer unack and broker send msg to onsumer occur concurrent?

such as :

time →
broker: |send to consumer & release lock|
client  : &nbsp;&nbsp;&nbsp; |start unack&wait the lock| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | get lock and do unack|

Consumer msgs: (1,2,3)
now unack 1, and broker send (4,5,6) to consumer at the same time after unack msg send to broker.

Will consumer consume the msg 4 before msg 1?

**Comment by asafm** [comment]
If order matters, I don't think you should negatively acknowledge a message in the Key-Shared subscription.

**Reply by xrayw** [reply]
@asafm  Thanks your reply

But if a message consume failed, we need to try again after a short period of time.

**Reply by asafm** [reply]
You then lose the ordering. If this is the case, why not use the Shared subscription type?

**Reply by xrayw** [reply]
😂 we need ensure the order of messages.

**Reply by merlimat** [reply, marked as answer]
If you do negative-ack, you are requesting a message to be redelivered later. In the meantime, you will receive other messages, potentially for the same key, therefore you will get out-of-order. 

If you want to ensure strict ordering, you need to pause and hold on to the message and keep trying to process the same message until you succeed, before resuming to the rest of messages.

**Reply by xrayw** [reply]
Got it. thanks @asafm @merlimat 👍
